File: platforms/windows/OneAI/Services/AppLocale.cs
// AppLocale — the Windows app's effective UI + engine language.
//
// Single source of truth read by both the group-chat engine locale (folded
// into the ScenarioSpecDto JSON → parse_scenario → ChatLocale) and the
// bilingual scenario presets (ScenarioStore.PresetsFor). The Settings panel's
// language picker sets `Override` (and, for chrome/.resw, the matching
// ApplicationLanguages.PrimaryLanguageOverride) — so a user override reaches
// the engine, the presets, and the localized chrome together. With no override
// set, the locale follows the system UI culture (which is also what .resw
// resource resolution uses by default).

using System.Globalization;

namespace OneAI.Services;

/// <summary>The app's effective language: zh or en.</summary>
public enum AppLocale
{
    Zh,
    En,
}

public static class AppLocaleHelper
{
    /// <summary>Explicit user override ("zh"/"en"), or null = follow system.
    /// Set by the Settings language picker (layer-2 chrome wiring).</summary>
    public static string? Override { get; set; }

    /// <summary>The effective locale: explicit override if set, otherwise the
    /// system UI culture. `zh` for any Chinese variant, `en` for everything
    /// else (the English fallback).</summary>
    public static AppLocale Current
    {
        get
        {
            if (Override is { } o)
            {
                if (o.StartsWith("zh", StringComparison.OrdinalIgnoreCase)) return AppLocale.Zh;
                if (o.StartsWith("en", StringComparison.OrdinalIgnoreCase)) return AppLocale.En;
            }
            string name = CultureInfo.CurrentUICulture.Name;
            return name.StartsWith("zh", StringComparison.OrdinalIgnoreCase) ? AppLocale.Zh : AppLocale.En;
        }
    }

    /// <summary>The c_facade `locale` JSON value ("en"/"zh").</summary>
    public static string? LocaleCode => Current switch
    {
        AppLocale.En => "en",
        _ => "zh",
    };
}

File: platforms/macos/Sources/AppLocale.swift
// AppLocale — the app's effective UI + engine language.
//
// Single source of truth read by both the group-chat engine locale
// (passed to ScenarioSpecView.locale → ChatLocaleView) and the bilingual
// scenario presets (AgentStore.presets(locale:)). The Settings panel's
// language picker writes the `oneai_language` UserDefaults key (and, for
// SwiftUI Text chrome, the matching `AppleLanguages` override) — so a user
// override reaches the engine, the presets, and the localized chrome together.
// With no override set, the locale follows the system preferred languages
// (which is also what SwiftUI Text + Bundle.main use by default).

import Foundation

enum AppLocale: String, Codable, CaseIterable {
    case zh, en

    /// The effective locale: explicit user override (`oneai_language`) if set,
    /// otherwise the system preferred language. `zh` for any Chinese variant,
    /// `en` for everything else (the English fallback).
    static var current: AppLocale {
        if let raw = UserDefaults.standard.string(forKey: "oneai_language"),
           let loc = AppLocale(rawValue: raw) {
            return loc
        }
        let preferred = Locale.preferredLanguages.first ?? ""
        return preferred.hasPrefix("zh") ? .zh : .en
    }

    /// The engine `ChatLocaleView` for `ScenarioSpecView.locale`.
    var chatLocaleView: ChatLocaleView {
        switch self {
        case .en: return .en
        case .zh: return .zh
        }
    }
}
